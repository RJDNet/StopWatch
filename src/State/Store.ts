import {
    Store,
    Unsubscribe
} from 'redux';

import { IState } from './IState';

export let store: Readonly<Store<IState>>;

export function getState(): Readonly<IState> {
    return store.getState();
}

export function setStore(storeToSet: Readonly<Store<IState>>): void {
    store = storeToSet;
}

export function dispatch(action: { type: string }): void {
    if (store !== undefined) {
        store.dispatch(action);
    }
}

export function subscribe(listener: () => void): Unsubscribe {
    return store.subscribe(listener);
}