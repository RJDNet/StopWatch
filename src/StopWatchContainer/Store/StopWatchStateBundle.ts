import { IRemoveStopWatch } from './StopWatchStateBundle';
import uuidv1 from 'uuid/v1';

// Action Types
export enum ActionTypes {
    ADD_STOPWATCH = 'ADD_STOPWATCH',
    ADDX_STOPWATCH = 'ADDX_STOPWATCH',
    REMOVEALL_STOPWATCH = 'REMOVEALL_STOPWATCH',
    REMOVE_STOPWATCH = 'REMOVE_STOPWATCH',
    START_STOPWATCH = 'START_STOPWATCH',
    STOP_STOPWATCH = 'STOP_STOPWATCH',
    TICK_STOPWATCH = 'TICK_STOPWATCH',
    RESET_STOPWATCH = 'RESET_STOPWATCH'
}

// Action Interfaces
export interface IAddStopWatch {
    type: ActionTypes.ADD_STOPWATCH;
}

export interface IAddXStopWatch {
    type: ActionTypes.ADDX_STOPWATCH;
    payload: {
        amount: number;
    }
}

export interface IRemoveAllStopWatch {
    type: ActionTypes.REMOVEALL_STOPWATCH;
}

export interface IRemoveStopWatch {
    type: ActionTypes.REMOVE_STOPWATCH;
    payload: {
        key: string;
    }
}

export interface IStartStopWatch {
    type: ActionTypes.START_STOPWATCH;
    payload: {
        key: string;
    }
}

export interface IStopStopWatch {
    type: ActionTypes.STOP_STOPWATCH;
    payload: {
        key: string;
    }
}

export interface ITickStopWatch {
    type: ActionTypes.TICK_STOPWATCH;
    payload: {
        key: string;
    }
}

export interface IResetStopWatch {
    type: ActionTypes.RESET_STOPWATCH;
    payload: {
        key: string;
    }
}

// Consolidate Action Interfaces
export type Actions = IAddStopWatch |
    IAddXStopWatch |
    IRemoveAllStopWatch |
    IRemoveStopWatch |
    IStartStopWatch |
    IStopStopWatch |
    ITickStopWatch |
    IResetStopWatch;

// Action Creators
export const actionCreators = {
    addWatch(): IAddStopWatch {
        const action: IAddStopWatch = {
            type: ActionTypes.ADD_STOPWATCH
        }
        
        return action;
    },
    addXWatch(amount: number): IAddXStopWatch {
        const action: IAddXStopWatch = {
            type: ActionTypes.ADDX_STOPWATCH,
            payload: {
                amount
            }
        }
        
        return action;
    },
    removeAllWatches(): IRemoveAllStopWatch {
        const action: IRemoveAllStopWatch = {
            type: ActionTypes.REMOVEALL_STOPWATCH
        }
        
        return action;
    },
    removeWatch(key: string): IRemoveStopWatch {
        const action: IRemoveStopWatch = {
            type: ActionTypes.REMOVE_STOPWATCH,
            payload: {
                key
            }
        }
        
        return action;
    },
    startWatch(key: string): IStartStopWatch {
        const action: IStartStopWatch = {
            type: ActionTypes.START_STOPWATCH,
            payload: {
                key
            }
        }
        
        return action;
    },
    stopWatch(key: string): IStopStopWatch {
        const action: IStopStopWatch = {
            type: ActionTypes.STOP_STOPWATCH,
            payload: {
                key
            }
        }
            
        return action;
    },
    tickWatch(key: string): ITickStopWatch {
        const action: ITickStopWatch = {
            type: ActionTypes.TICK_STOPWATCH,
            payload: {
                key
            }
        }
            
        return action;
    },
    resetWatch(key: string): IResetStopWatch {
        const action: IResetStopWatch = {
            type: ActionTypes.RESET_STOPWATCH,
            payload: {
                key
            }
        }
        
        return action;
    }
}

export interface IStopWatch {
    time: number;
    running: boolean;
    key: string;
}

export interface IStopWatchMap {
    [key: string]: IStopWatch;
}

export interface IStopWatchState {
    stopWatches: IStopWatchMap;
    stopWatchList: string[];
}

// Default value of store
export function getInitialStopWatchState(): Readonly<IStopWatchState> {
    const state: Readonly<IStopWatchState> = {
        stopWatches: {},
        stopWatchList: []
    }

    return state;
}

// Sub Reducers
export function addWatch(state: IStopWatchState): IStopWatchState {
    const key: string = uuidv1();

    const update: IStopWatchState = {
        ...state,
        stopWatches: {
            ...state.stopWatches,
            [key]: {
                time: 0,
                running: false,
                key
            }
        },
        stopWatchList: [...state.stopWatchList, key]
    }

    return update;
}

export function addXWatch(state: IStopWatchState, action: IAddXStopWatch): IStopWatchState {
    const { amount } = action.payload;
    let update = state;
    
    for(let i = 1; i <= amount; i++) {
        const key: string = uuidv1();
        let newUpdate: IStopWatchState = {
            ...update,
            stopWatches: {
                ...update.stopWatches,
                [key]: {
                    time: 0,
                    running: false,
                    key
                }
            },
            stopWatchList: [...update.stopWatchList, key]
        }

        update = newUpdate;
    }
    

    return update;
}

function removeAllWatches(state: IStopWatchState): IStopWatchState {
    const update: IStopWatchState = {
        ...state,
        stopWatches: {},
        stopWatchList: []
    }

    return update;
}

function removeWatch(state: IStopWatchState, action: IRemoveStopWatch): IStopWatchState {
    const { key } = action.payload;
    const delKey = {
        ...state
    }
    delete delKey.stopWatches[key]

    const update: IStopWatchState = {
        ...state,
        stopWatches: {
            ...delKey.stopWatches
        },
        stopWatchList: state.stopWatchList.filter(f => f !== key)
    }

    return update;
}

function startWatch(state: IStopWatchState, action: IStartStopWatch): IStopWatchState {
    const { key } = action.payload;

    const update: IStopWatchState = {
        ...state,
        stopWatches: {
            ...state.stopWatches,
            [key]: {
                ...state.stopWatches[key],
                running: true
            }
        }
    }

    return update;
}

function stopWatch(state: IStopWatchState, action: IStopStopWatch): IStopWatchState {
    const { key } = action.payload;

    const update: IStopWatchState = {
        ...state,
        stopWatches: {
            ...state.stopWatches,
            [key]: {
                ...state.stopWatches[key],
                running: false
            }
        }
    }

    return update;
}

function tickWatch(state: IStopWatchState, action: ITickStopWatch): IStopWatchState {
    const { key } = action.payload;

    if (state.stopWatches[key] && state.stopWatches[key].running) {
        const time: number = state.stopWatches[key].time + 1;
        const update: IStopWatchState = {
            ...state,
            stopWatches: {
                ...state.stopWatches,
                [key]: {
                    ...state.stopWatches[key],
                    time
                }
            }
        }
    
        return update;
    }

    return state;
}

function resetWatch(state: IStopWatchState, action: IResetStopWatch): IStopWatchState {
    const { key } = action.payload;

    const update: IStopWatchState = {
        ...state,
        stopWatches: {
            ...state.stopWatches,
            [key]: {
                ...state.stopWatches[key],
                running: false,
                time: 0
            }
        }
    }

    return update;
}

// Reducer
export const stopWatchReducer = (state: IStopWatchState = getInitialStopWatchState(), action: Actions ) => {
    switch (action.type) {
      case ActionTypes.ADD_STOPWATCH:
        return addWatch(state);
      case ActionTypes.ADDX_STOPWATCH:
        return addXWatch(state, action);
      case ActionTypes.REMOVEALL_STOPWATCH:
        return removeAllWatches(state);
      case ActionTypes.REMOVE_STOPWATCH:
        return removeWatch(state, action);
      case ActionTypes.START_STOPWATCH: 
        return startWatch(state, action);
      case ActionTypes.STOP_STOPWATCH: 
        return stopWatch(state, action);
      case ActionTypes.TICK_STOPWATCH: 
        return tickWatch(state, action);
      case ActionTypes.RESET_STOPWATCH: 
        return resetWatch(state, action);
      default: return state;
    }
}

// Helpers
// function uuidv4() {
//     return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
//       var r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
//       return v.toString(16);
//     });
//   }

// export function uuidv4() {
//     var dt = new Date().getTime();
//     var uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
//         var r = (dt + Math.random()*16)%16 | 0;
//         dt = Math.floor(dt/16);
//         return (c==='x' ? r :(r&0x3|0x8)).toString(16);
//     });
//     return uuid;
// }
