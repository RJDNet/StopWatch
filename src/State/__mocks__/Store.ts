import { 
    createStore, 
    Store, 
    combineReducers,
    Reducer
  } from 'redux';

import { IState } from '../IState';
import {
    IStopWatchState,
    getInitialStopWatchState
} from '../../StopWatchContainer/Store/StopWatchStateBundle';
import { getReducers } from '../../State/GetReducers';

export function initialStoreState(): Readonly<IState> {
    const stopWatchState: IStopWatchState = { ...getInitialStopWatchState() };

    return {
        stopWatchState
    };
}

export function getState(): Readonly<IState> {
    const state: Readonly<IState> = initialStoreState();

    return { 
        ...state 
    };
}

export const dispatch = jest.fn();
export const subscribe = jest.fn();

export function setupStore(initialState: IState = getState()): Readonly<Store<IState>> {
    const reducer: Reducer<{}> = combineReducers(getReducers());

    return createStore(
        reducer,
        initialState,
        window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
    ) as Store<IState>;
}
  
export function getMockState(): Readonly<IState> {
    return store.getState();
}

export let store: any = setupStore();

function setStore(storeToSet: Readonly<Store<IState>>): void {
    store = storeToSet;
}

setStore(store);
getMockState();
  