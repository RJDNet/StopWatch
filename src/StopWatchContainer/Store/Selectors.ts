import { IState } from "../../State/IState";
import { IStopWatch } from "./StopWatchStateBundle";

export const getStopWatchList = (state: IState): string[] => {
    return state.stopWatchState.filteredList;
};

export const getStopWatch = (id: string, state: IState): IStopWatch | undefined => {
    if (state.stopWatchState.stopWatches[id]) {
        return state.stopWatchState.stopWatches[id];
    }
    
    return undefined;
};