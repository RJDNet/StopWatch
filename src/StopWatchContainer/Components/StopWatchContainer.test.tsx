import React from 'react';
import { Provider } from "react-redux";
import { 
    shallow, 
    ShallowWrapper 
} from 'enzyme';
import toJson from 'enzyme-to-json';

import { 
    store,
    getState 
} from '../../State/__mocks__/Store';
import { ActionTypes } from '../Store/StopWatchStateBundle';
import ConnectedStopWatchContainer, { 
    StopWatchContainer, 
    mapStateToProps 
} from './StopWatchContainer';

describe('StopWatchContainer Component', () => {
    const dispatch = jest.fn();
    const props = {
        dispatch,
        stopWatchList: []
    }

    test('StopWatch renders correctly with all props & no redux store', () => {        
        const tree: ShallowWrapper = shallow(
          <StopWatchContainer {...props} />
        );
          
        expect(toJson(tree)).toMatchSnapshot();
      });

    test('StopWatchContainer renders correctly with no props', () => {
        const component: ShallowWrapper = shallow(
            <Provider store={store}>
                <ConnectedStopWatchContainer />
            </Provider>
        );
      
        expect(toJson(component)).toMatchSnapshot();
    });

    test('mapStateToProps returns expected state', () => {
        const state = getState();
        const result = mapStateToProps(state);

        expect(result).toBeDefined();

        expect(result.stopWatchList).toEqual([]);
    });

    test('button click should add stopwatch', () => {
        const tree: ShallowWrapper = shallow(
            <StopWatchContainer {...props} />
        );
    
        tree
          .find('button.stopwatches-container__button-container__add-button')
          .simulate('click');
            expect(dispatch).toHaveBeenCalled();
            expect(dispatch).toBeCalledWith({
                type: ActionTypes.ADD_STOPWATCH
            });
    });

    test('button click should add x stopwatches', () => {
        const tree: ShallowWrapper = shallow(
            <StopWatchContainer {...props} />
        );

        const testvalue: number = 5;
        tree.find('input').simulate('change', {target: {value: testvalue}});
        tree.update();
    
        tree
          .find('button.stopwatches-container__button-container__addx-button')
          .simulate('click');
            expect(dispatch).toHaveBeenCalled();
            expect(dispatch).toBeCalledWith({
                type: ActionTypes.ADDX_STOPWATCH,
                payload: {
                    amount: testvalue
                }
            });
    });

    test('button click should remove all stopwatches', () => {
        const tree: ShallowWrapper = shallow(
            <StopWatchContainer {...props} />
        );
    
        tree
          .find('button.stopwatches-container__button-container__removeall-button')
          .simulate('click');
            expect(dispatch).toHaveBeenCalled();
            expect(dispatch).toBeCalledWith({
                type: ActionTypes.REMOVEALL_STOPWATCH
            });
    });
});