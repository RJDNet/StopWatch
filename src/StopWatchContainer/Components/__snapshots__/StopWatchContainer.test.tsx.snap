// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopWatchContainer Component StopWatch renders correctly with all props & no redux store 1`] = `
<div
  className="stopwatches-container"
>
  <h1
    className="stopwatches-container__header"
  >
    Stopwatch App
  </h1>
  <div
    className="stopwatches-container__controls-container"
  >
    <p
      className="stopwatches-container__amount"
    >
      Stopwatch Amount: 
      0
    </p>
    <div
      className="stopwatches-container__button-container"
    >
      <button
        className="stopwatches-container__button-container__add-button"
        onClick={[Function]}
      >
        ADD ONE
      </button>
      <button
        className="stopwatches-container__button-container__removeall-button"
        onClick={[Function]}
      >
        REMOVE ALL
      </button>
      <div
        className="stopwatches-container__button-container__dropdown-container"
      >
        <button
          className="stopwatches-container__button-container__dropdown-container__dropdown-button"
        >
          Filter By â–¼
        </button>
        <div
          className="stopwatches-container__button-container__dropdown-container__dropdown-content"
        >
          <span
            onClick={[Function]}
          >
            All
          </span>
          <span
            onClick={[Function]}
          >
            Running
          </span>
          <span
            onClick={[Function]}
          >
            Not Running
          </span>
        </div>
      </div>
    </div>
    <div
      className="stopwatches-container__button-container"
    >
      <input
        className="stopwatches-container__button-container__amount-input"
        onChange={[Function]}
        placeholder="Enter X Amount..."
        type="number"
        value=""
      />
      <button
        className="stopwatches-container__button-container__addx-button"
        onClick={[Function]}
      >
        ADD X
      </button>
    </div>
  </div>
  <div
    className="stopwatches-container__stopwatch-flexcontainer"
  />
</div>
`;

exports[`StopWatchContainer Component StopWatchContainer renders correctly with no props 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <Connect(StopWatchContainer) />
</ContextProvider>
`;
